app/controllers/predicates/actions/AuthorisedAction.scala

```diff
@@ -106,53 +107,78 @@ class AuthorisedAction @Inject() (appConfig: AppConfig)(implicit val authService
-                  logger.info(s"[AuthorisedAction][agentAuthentication] - No session id in request")
-                  Future(Redirect(appConfig.signInUrl))
-                } { sessionId =>
-                  block(AuthorisationRequest(User(mtdItId, Some(arn), nino, sessionId, AffinityGroup.Agent.toString), request))
-                }
-
-              case None =>
-                logger.info("[AuthorisedAction][agentAuthentication] Agent with no HMRC-AS-AGENT enrolment. Rendering unauthorised view.")
-                Future.successful(Redirect(controllers.errors.routes.YouNeedAgentServicesController.show))
-            }
-          } recover {
-          case _: NoActiveSession =>
-            logger.info(s"[AuthorisedAction][agentAuthentication] - No active session. Redirecting to ${appConfig.signInUrl}")
-            Redirect(appConfig.signInUrl)
-          case ex: AuthorisationException =>
-            logger.info(s"[AuthorisedAction][agentAuthentication] - Agent does not have delegated authority for Client.")
-            Redirect(controllers.errors.routes.AgentAuthErrorController.show)
-        }
+          .retrieve(allEnrolments) {
+            populateAgent(block, mtdItId, nino, _, AffinityGroup.Agent, isSecondaryAgent = false)
+          }
+          .recoverWith(agentRecovery(block, mtdItId, nino, AffinityGroup.Agent))
+
       case (mtditid, nino) =>
         logger.info(
-          s"[AuthorisedAction][agentAuthentication] - Agent does not have session key values. " +
-            s"Redirecting to view & change. MTDITID missing:${mtditid.isEmpty}, NINO missing:${nino.isEmpty}")
+          s"[AuthorisedAction][agentAuthentication] - Agent does not session key values. Redirecting to view & change." +
+            s"MTDITID missing:${mtditid.isEmpty}, NINO missing:${nino.isEmpty}")
         Future.successful(Redirect(appConfig.viewAndChangeEnterUtrUrl))
     }
+
+  private def agentRecovery[A](block: AuthorisationRequest[A] => Future[Result], mtdItId: String, nino: String, affinityGroup: AffinityGroup)(implicit
+      request: Request[A],
+      hc: HeaderCarrier): PartialFunction[Throwable, Future[Result]] = {
+    case _: NoActiveSession =>
+      logger.info(s"[AuthorisedAction][agentAuthentication] - No active session. Redirecting to ${appConfig.signInUrl}")
+      Future(Redirect(appConfig.signInUrl))
+    case _: AuthorisationException =>
+      if (appConfig.emaSupportingAgentsEnabled) {
+        authService
+          .authorised(secondaryAgentPredicate(mtdItId))
+          .retrieve(allEnrolments) {
+            populateAgent(block, mtdItId, nino, _, affinityGroup, isSecondaryAgent = true)
+          }
+          .recoverWith {
+            case _: AuthorisationException =>
+              logger.info(s"[AuthorisedAction][agentAuthentication] - Agent does not have secondary delegated authority for Client.")
+              Future(Redirect(controllers.errors.routes.AgentAuthErrorController.show))
+            case _ =>
```

Comment:
See this PR here for additional logic required:

https://github.com/hmrc/income-tax-expenses/pull/28